source/extensions/filters/common/fault/fault_config.cc

```diff
@@ -11,24 +11,54 @@ namespace Filters {
 namespace Common {
 namespace Fault {
 
+envoy::type::v3::FractionalPercent
+HeaderPercentageProvider::percentage(const Http::RequestHeaderMap* request_headers) const {
+  if (!request_headers) {
+    return percentage_;
+  }
+
+  const auto header = request_headers->get(header_name_);
+  if (header == nullptr) {
+    return percentage_;
+  }
+
+  uint64_t header_numerator;
+  if (!absl::SimpleAtoi(header->value().getStringView(), &header_numerator)) {
+    return percentage_;
+  }
+
+  if (header_numerator > 100) {
+    return percentage_;
+  }
+
+  envoy::type::v3::FractionalPercent header_percentage;
+  header_percentage.set_numerator(header_numerator);
+  return header_percentage;
+}
+
 FaultAbortConfig::FaultAbortConfig(
-    const envoy::extensions::filters::http::fault::v3::FaultAbort& abort_config)
-    : percentage_(abort_config.percentage()) {
+    const envoy::extensions::filters::http::fault::v3::FaultAbort& abort_config) {
   switch (abort_config.error_type_case()) {
   case envoy::extensions::filters::http::fault::v3::FaultAbort::ErrorTypeCase::kHttpStatus:
-    provider_ = std::make_unique<FixedAbortProvider>(abort_config.http_status());
+    provider_ =
+        std::make_unique<FixedAbortProvider>(abort_config.http_status(), abort_config.percentage());
     break;
   case envoy::extensions::filters::http::fault::v3::FaultAbort::ErrorTypeCase::kHeaderAbort:
-    provider_ = std::make_unique<HeaderAbortProvider>();
+    provider_ = std::make_unique<HeaderAbortProvider>(abort_config.percentage());
     break;
   case envoy::extensions::filters::http::fault::v3::FaultAbort::ErrorTypeCase::ERROR_TYPE_NOT_SET:
     NOT_REACHED_GCOVR_EXCL_LINE;
   }
 }
 
-absl::optional<Http::Code>
-FaultAbortConfig::HeaderAbortProvider::statusCode(const Http::HeaderEntry* header) const {
+absl::optional<Http::Code> FaultAbortConfig::HeaderAbortProvider::statusCode(
+    const Http::RequestHeaderMap* request_headers) const {
   absl::optional<Http::Code> ret;
+  if (!request_headers) {
+    return absl::nullopt;
+  }
```

Comment:
I don't think this can happen? Same elsehwere.

Reply:
Yes, it will never happen. I added this a safety measure in case we start using this class in more places in the future. That being said, I can remove this check if this kind of defensive programming is not something we do in Envoy repo.

Reply:
Yes in general we don't put in this type of defensive logic which we know we can't hit currently, so I would remove.

Reply:
Just so that we are on the same page. There is one place where we call `percentage` method with empty headers (`nullptr`) in Mongo related code but I don't think that header faults are supported there so we should be fine. The code I'm talking about can be found here https://github.com/envoyproxy/envoy/pull/10724/files#diff-eb7cf1a112138011ff766f2a9dc402f8R405